### PR TITLE
Display credentials report

### PR DESCRIPTION
**Pull request by abdusahin**
- Improved UI

### CODE REVIEW

File: hq/app/logic/ReportDisplay.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+  private[logic] def key1Status(cred: IAMCredential): KeyStatus = if (cred.accessKey1Active) AccessKeyEnabled else if (isKey1Disabled(cred)) AccessKeyDisabled else NoKey
+  private[logic] def key2Status(cred: IAMCredential): KeyStatus = if (cred.accessKey2Active) AccessKeyEnabled else if (isKey2Disabled(cred)) AccessKeyDisabled else NoKey
+
+  private[logic] def machineReportStatus(cred: IAMCredential): ReportStatus = {
+
+    if (!Seq(key1Status(cred), key2Status(cred)).contains(AccessKeyEnabled)) Amber else Green
+  }
+
+  private[logic] def humanReportStatus(cred: IAMCredential): ReportStatus =
```
Comment (adamnfish) [inline hq/app/logic/ReportDisplay.scala]: We've got a heady mix of if statements spread over heaps of lines and some squeezed onto one. For my money, the options in Scala are:

Simple if statement:

    if (something) a else b

Medium complexity if statement:

    if (something a bit more complex) slight more complex a
    else slightly more complex b

Complex if statements:

    if (something) {
      // some logic in here
    } else {
      // logic
    }

Here we have similar if statements with very different displays, and this one with the values on new lines and no braces is a bit unusual.

This is a minor point, but it'd be good to settle on something here.

File: hq/public/stylesheets/main.css [hunk 2]
```diff
@@ -184,3 +184,15 @@ nav {
 .collapsible-body {
     background-color: #ffffff;
 }
+
+i.key-disabled {
+    opacity:0.3
```
Comment (adamnfish) [inline hq/public/stylesheets/main.css]: `opacity: 0.3;`

File: hq/app/logic/ReportDisplay.scala [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+object ReportDisplay {
+
+
+  private[logic] def lastActivityDate(cred: IAMCredential): Option[DateTime] = {
+    val allDates =
+      cred.passwordLastUsed.toSeq ++ cred.accessKey1LastUsedDate.toSeq  ++ cred.accessKey2LastUsedDate.toSeq
+    allDates.sortWith(_.isAfter(_)).collectFirst { case date if date.isBefore(DateTime.now(DateTimeZone.UTC)) => date }
+  }
+
+  private def isKey1Disabled(cred: IAMCredential): Boolean = (!cred.accessKey1Active) && cred.accessKey1LastUsedDate.toSeq.nonEmpty
```
[the remark this answers is not in the document]
Comment (abdusahin) [reply hq/app/logic/ReportDisplay.scala]: 👍

File: hq/app/logic/ReportDisplay.scala [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  private def isKey2Disabled(cred: IAMCredential): Boolean = (!cred.accessKey2Active) && cred.accessKey2LastUsedDate.toSeq.nonEmpty
+
+  private[logic] def key1Status(cred: IAMCredential): KeyStatus = if (cred.accessKey1Active) AccessKeyEnabled else if (isKey1Disabled(cred)) AccessKeyDisabled else NoKey
```
[the remark this answers is not in the document]
Comment (abdusahin) [reply hq/app/logic/ReportDisplay.scala]: merged with isKeyDisabled

File: hq/app/logic/ReportDisplay.scala [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+          )
+        else None
+
+      val humanUser =
+        if (cred.passwordEnabled.getOrElse(false)) {
+          Some(HumanUser(cred.user, cred.mfaActive, key1Status(cred), key2Status(cred), humanReportStatus(cred), DateUtils.dayDifference(lastActivityDate(cred))))
+        }
+        else None
+      report.copy(
+        machineUsers = report.machineUsers ++ machineUser.toSeq, humanUsers = report.humanUsers ++ humanUser.toSeq)
```
[the remark this answers is not in the document]
Comment (abdusahin) [reply hq/app/logic/ReportDisplay.scala]: done

File: hq/app/views/iam/iamCredReport.scala.html [hunk 6]
```diff
@@ -1,5 +1,93 @@
...
+@import model.AccessKeyEnabled
+@import logic.DateUtils
+@import model.KeyStatus
+@import model.ReportStatus
+@(report: CredentialReportDisplay)
+
+
+
+@keyStatus(keyStatus: KeyStatus)  = {
+    @{
```
[the remark this answers is not in the document]
Comment (abdusahin) [reply hq/app/views/iam/iamCredReport.scala.html]: this style isn't supported by twirl version 1.1.1, it's supported version  1.2.1 onwards.

File: hq/test/logic/ReportDisplayTest.scala [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+      val machineCred = cred.copy(passwordEnabled = None, passwordLastUsed = None, accessKey2Active = true, accessKey1Active = false)
+      ReportDisplay.machineReportStatus(machineCred) shouldBe Green
+    }
+
+    "check machine report status amber when no key enabled" in {
+      val machineCred = cred.copy(passwordEnabled = None, passwordLastUsed = None, accessKey1Active = false, accessKey2Active = false)
+      ReportDisplay.machineReportStatus(machineCred) shouldBe Amber
+    }
+
+    "check human report status green when mfa active and no active access key" in {
```
[the remark this answers is not in the document]
Comment (abdusahin) [reply hq/test/logic/ReportDisplayTest.scala]: done

**Comment by abdusahin** [comment]
@adamnfish can you have another look.

**Comment by adamnfish** [comment]
Minor points now :-)

Sort the Twirl template (either make it Twirl-first or move it to a Scala function) and then :+1:

Looks great!

File: hq/app/logic/ReportDisplay.scala [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  private[logic] def humanReportStatus(cred: IAMCredential): ReportStatus = {
+    if (!cred.mfaActive)
+      Red
+    else if (Seq(key1Status(cred), key2Status(cred)).contains(AccessKeyEnabled))
+      Amber
+    else Green
+  }
+
+  def toCredentialReportDisplay(report: IAMCredentialsReport): CredentialReportDisplay = {
+    import DateUtils._
```
[the remark this answers is not in the document]
Comment (abdusahin) [reply hq/app/logic/ReportDisplay.scala]: not used any where else, but don't mind other way either.

**Comment by abdusahin** [comment]
@adamnfish  , i updated according to the feedback, can you have a look.

**Comment by adamnfish** [comment]
:+1:

**Pull request merged**